spacy/pipeline/entity_linker.py

```diff
@@ -451,132 +445,127 @@ def predict(self, docs: Iterable[Doc]) -> ActivationsT:
             }
         if isinstance(docs, Doc):
             docs = [docs]
-        for doc in docs:
+
+        docs_iters = tee(docs, 2)
+
+        # Call candidate generator.
+        all_ent_cands = self.get_candidates(
+            self.kb,
+            (
+                SpanGroup(
+                    doc,
+                    spans=[
+                        ent for ent in doc.ents if ent.label_ not in self.labels_discard
+                    ],
+                )
+                for doc in docs_iters[0]
+            ),
+        )
+
+        for doc in docs_iters[1]:
             doc_ents: List[Ints1d] = []
             doc_scores: List[Floats1d] = []
-            if len(doc) == 0:
+            if len(doc) == 0 or len(doc.ents) == 0:
                 docs_scores.append(Ragged(ops.alloc1f(0), ops.alloc1i(0)))
                 docs_ents.append(Ragged(xp.zeros(0, dtype="uint64"), ops.alloc1i(0)))
                 continue
             sentences = [s for s in doc.sents]
-
-            # Loop over entities in batches.
-            for ent_idx in range(0, len(doc.ents), self.candidates_batch_size):
-                ent_batch = doc.ents[ent_idx : ent_idx + self.candidates_batch_size]
-
-                # Look up candidate entities.
-                valid_ent_idx = [
-                    idx
-                    for idx in range(len(ent_batch))
-                    if ent_batch[idx].label_ not in self.labels_discard
-                ]
-
-                batch_candidates = list(
-                    self.get_candidates_batch(
-                        self.kb,
-                        SpanGroup(doc, spans=[ent_batch[idx] for idx in valid_ent_idx]),
-                    )
-                    if self.candidates_batch_size > 1
-                    else [
-                        self.get_candidates(self.kb, ent_batch[idx])
-                        for idx in valid_ent_idx
-                    ]
+            doc_ent_cands = list(next(all_ent_cands))
+
+            # Looping over candidate entities for this doc. (TODO: rewrite)
+            for ent_cand_idx, ent in enumerate(doc.ents):
```

Comment:
Does our test suite sufficiently cover an execution path on this loop, just in case anything's gone wrong in the refactor?

Reply:
Not sure what you mean, can you elaborate? All tests dealing with entities run this loop (the outer loop is new and running over (candidates per) docs, the inner loop runs over the candidates for a single doc).

Reply:
Added one in https://github.com/explosion/spaCy/pull/12367/commits/eef3de098f3d422e19c7583c47f66d4a4581941c.